src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1537,4 +1536,105 @@ impl Cage {
+                                let mut vec = vec![0u8; length];
+                                let buf: *mut u8 = &mut vec[0];
+                                
+                                // read the part to keep to properly sized buffer
+                                if let Ok(bytesread) = fileobject.readat(buf, length, 0) {
+                                    // remove file from file object table if readat() succeeded
+                                    fobjtable.remove(&inodenum).unwrap().close().unwrap();
+                                    let filename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    interface::removefile(filename).unwrap();
+
+                                    // insert new file created using old filename to file object table
+                                    let filename1 = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    fobjtable.insert(inodenum, interface::openfile(filename1, true).unwrap());
+                                    fileobject = fobjtable.get_mut(&inodenum).unwrap();
+
+                                    // write the kept part to new file from buffer
+                                    if let Ok(byteswritten) = fileobject.writeat(buf, length, 0) {
+                                        // update filesize if writeat() succeeded
+                                        normalfile_inode_obj.size = length;
+                                    } else {
+                                        panic!("writeat() failed");
+                                    }
+                                } else {
+                                   panic!("readat() failed");
+                                }
+                            } 
+                        }
+                        Inode::CharDev(_) => {
+                            return syscall_error(Errno::EISDIR, "ftruncate", "The named file is a character driver");
+                        }
+                        Inode::Dir(_) => {
+                            return syscall_error(Errno::EISDIR, "ftruncate", "The named file is a directory");
+                        }
+                        _ => {
+                            panic!("A file fd points to a socket or pipe");
+                        }
+                    };
+                }
+                _ => {
+                    return syscall_error(Errno::EINVAL, "ftruncate", "fd does not reference a regular file");
+                }
+            };
+            0 // ftruncate() has succeeded!
+        } else { 
+            syscall_error(Errno::EBADF, "ftruncate", "fd is not a valid file descriptor")
+        }
+    }
+
+    //------------------TRUNCATE SYSCALL------------------
+    pub fn truncate_syscall(&self, path: &str, length: usize) -> i32 {
+        self.ftruncate_syscall(self.open_syscall(path, O_RDWR, S_IRWXA), length)
+    }
```

Comment:
Doing it like this is not really what we want, it seems to have been included as a quick hack in the old safeposix. It would be better if you could do it like stat/fstat where you get the inode number of the file in the one that takes the fd from within the fd, and the inode number of the file in the one that takes the file path using metawalk, and then call from both into a shared helper function. It should be pretty straightforward.

Reply:
It's actually fine to keep it this way, because we need an open fobj to do trunc.